### ISSUE TITLE
An error occurred during deployment of openstack::HostPort exception: KeyError

### ISSUE THREAD

**Issue by automate-solutions** [inmanta/openstack#2]
When trying to deploy an instance on Openstack, I have the following errors:
```
[root@inmanta openstack-test]# inmanta -vvv deploy
INFO     Creating new virtual environment in ./.env
DEBUG    Storing state data in ./data/deploy
DEBUG    Starting mongodb on port 40716
INFO     Starting server endpoint
INFO     Connected to mongodb database inmanta on localhost:40716
WARNING  Config name shared-secret not defined in section server
DEBUG    Cleaning server session DB
DEBUG    Started server on port 41973
INFO     iaas_openstack matches agents managed by server, ensuring it is started.
ssh: Could not resolve hostname iaas_openstack: Name or service not known
DEBUG    New session for agent iaas_openstack on inmanta
DEBUG    set session 96e8a924-5ccc-11e7-a65b-000c2928d570 as primary for agent iaas_openstack in env 54850f7b-8e23-475a-a7de-d34ca98c6a80
DEBUG    Started new agent with PID 2281
INFO     iaas_openstack matches agents managed by server, ensuring it is started.
503 GET /parameter/ip_address?resource_id=openstack%3A%3AVirtualMachine%5Biaas_openstack%2Cname%3Dtesthost%5D (::1) 13.49ms
DEBUG    Successfully stored version 1498741944
INFO     ['iaas_openstack'] matches agents managed by server, ensuring it is started.
[0 / 2]g deploy
DEBUG    Updating/setting parameter subnet_private_ip_0 in env 54850f7b-8e23-475a-a7de-d34ca98c6a80 (for resource openstack::VirtualMachine[iaas_openstack,name=testhost])
openstack::VirtualMachine[iaas_openstack,name=testhost],v=1498741944 - deployed-a7de-d34ca98c6a80 (for resource openstack::VirtualMachine[iaas_openstack,name=testhost])
openstack::HostPort[iaas_openstack,name=testhost_eth0],v=1498741944 - failed 
[2 / 2]
```
The following Inmanta packages are installed on a Fedora 25:
```
[root@inmanta openstack-test]# dnf list python3-inmanta python3-inmanta-server python3-inmanta-agent
Last metadata expiration check: 1:16:20 ago on Thu Jun 29 14:18:39 2017.
Installed Packages
python3-inmanta.x86_64                                                                2017.1.2-1.fc25                                                         @bartvanbrabant-inmanta
python3-inmanta-agent.x86_64                                                          2017.1.2-1.fc25                                                         @bartvanbrabant-inmanta
python3-inmanta-server.x86_64                                                         2017.1.2-1.fc25                                                         @bartvanbrabant-inmanta
```

main.cf:
```
import ssh
import redhat
import ubuntu
import openstack

## Edit this parameters
image_id = "52a7e16f-8d48-4369-af78-2da1d23d2011"
network_name = "private"
subnet_name = "private_subnet"
network_address = "10.0.0.0/24"

flavor_name = "m1.little"
ssh_public_key="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC7KC7uZ/31S2AfbFZ+3EL5xL29/AF6N5C5oFLDg0qxJwnbLMY9MmU/QbSYG13YkSIyujEGrOqgJH9Q3S6biQ3ibU2RfmXCm0lnWweayz9eKaBbHMF1CqsYnEe6Mt7r2NeOh2uxeJyko+tIeEvOvwC/JZfVV3tBPej0se20SShG/EPHQ2KjWRMMbDBZ57OZ8MT+xmg5qsWobKJHwoIzFYPwlQ07s0Qe7KGBFxBOp5vogZ9Y162Gas7M2/8uz/YfmaL+JIry5Fl88bBtirIZKro7LOq1pDernoqbHnheJo6PAiFzmyEqYVSG5Vlom6F7Gr0Rgmg4BR525q/ilj5sIC7t root@inmanta"

os = redhat::fedora23
## End edit

# register ssh key
ssh_key = ssh::Key(name="mykey", public_key=ssh_public_key)

# Define the OpenStack provider to use
provider = openstack::Provider(name="iaas_openstack", connection_url=std::get_env("OS_AUTH_URL"),
                               username=std::get_env("OS_USERNAME"),
                               password=std::get_env("OS_PASSWORD"),
                               tenant=std::get_env("OS_PROJECT_NAME"))

# Define the project/tenant to boot the VM in, but do not let inmanta manage it
project = openstack::Project(provider=provider, name=provider.tenant, description="", enabled=true,
                             managed=false)

# Define the network objects to connect the virtual machine to but again, do not manage them
net = openstack::Network(provider=provider, project=project, name=network_name, managed=false)
subnet = openstack::Subnet(provider=provider, project=project, network=net, dhcp=true, managed=false,
                           name=subnet_name, network_address=network_address)

# Define the virtual machine
vm = openstack::Host(provider=provider, project=project, key_pair=ssh_key, name="testhost",
                     image=image_id, os=os, flavor=flavor_name, user_data="", subnet=subnet)
```

project.yml:
```
name: snippet test
modulepath: [libs,/tmp/tmpmqtlmkgr]
downloadpath: /tmp/tmpmqtlmkgr
version: 1.0
repo: ['https://github.com/inmanta/']
requires:
        - openstack~=2.1.0
        - ip~=0.6.1
        - net~=0.4
        - redhat~=0.6.1
        - ssh~=0.3
        - std~=0.12.2

```
We can see the error in the agent-xxxx.err logs:
```
2017-06-29 15:12:26,990 ERROR   An error occurred during deployment of openstack::HostPort[iaas_openstack,name=testhost_eth0],v=1498741944 (exception: KeyError('port_security_enabled',))
```
I could pin point the issue. In `openstack/plugins/__init__.py` at line 1250:
`resource.portsecurity = port["port_security_enabled"]`
There's no safeguard used to retrieve `port_security_enabled` value in the port dictionary.

I clearly see, if I log the content of `port` that `port_security_enabled` is absent of it:
```
{'revision_number': 11, 'updated_at': '2017-06-29T13:09:08Z', 'id': '1e748bbd-bf68-489f-80d4-caf63378776c', 'status': 'ACTIVE', 'device_id': '289a6564-1fb8-4f3b-a0d1-d4a2a90f8421', 'mac_address': 'fa:16:3e:07:a0:13', 'allowed_address_pairs': [], 'description': '', 'project_id': '5878aebf371d4df3aa5bcb0cc6456720', 'name': '', 'security_groups': [], 'tags': [], 'tenant_id': '5878aebf371d4df3aa5bcb0cc6456720', 'network_id': '492d3953-00f7-4350-aee0-c7b5f35b049c', 'created_at': '2017-06-29T13:06:57Z', 'binding:vnic_type': 'normal', 'admin_state_up': True, 'device_owner': 'compute:None', 'fixed_ips': [{'subnet_id': '73bc057e-6b59-4f21-9b90-cd4cc36722b9', 'ip_address': '10.0.0.10'}], 'extra_dhcp_opts': []}
```
Wouldn't be better to use something like:
`resource.portsecurity = port.get("port_security_enabled","a_default_value")`

**Comment by bartv** [comment]
Could you provide me with the output of "neutron ext-list"? My guess is that this extension is not enabled on the openstack you are using. Until now, we did run into openstack deployments without.

We should indeed check if the attribute is available, but I will also have to check what the impact is on other things such as securitygroups.

**Comment by automate-solutions** [comment]
```
+---------------------------+-------------------------------------------------------------+
| alias                     | name                                                        |
+---------------------------+-------------------------------------------------------------+
| default-subnetpools       | Default Subnetpools                                         |
| network-ip-availability   | Network IP Availability                                     |
| network_availability_zone | Network Availability Zone                                   |
| auto-allocated-topology   | Auto Allocated Topology Services                            |
| ext-gw-mode               | Neutron L3 Configurable external gateway mode               |
| binding                   | Port Binding                                                |
| metering                  | Neutron Metering                                            |
| agent                     | agent                                                       |
| subnet_allocation         | Subnet Allocation                                           |
| l3_agent_scheduler        | L3 Agent Scheduler                                          |
| tag                       | Tag support                                                 |
| external-net              | Neutron external network                                    |
| flavors                   | Neutron Service Flavors                                     |
| net-mtu                   | Network MTU                                                 |
| availability_zone         | Availability Zone                                           |
| quotas                    | Quota management support                                    |
| l3-ha                     | HA Router extension                                         |
| provider                  | Provider Network                                            |
| multi-provider            | Multi Provider Network                                      |
| address-scope             | Address scope                                               |
| extraroute                | Neutron Extra Route                                         |
| subnet-service-types      | Subnet service types                                        |
| standard-attr-timestamp   | Resource timestamps                                         |
| service-type              | Neutron Service Type Management                             |
| l3-flavors                | Router Flavor Extension                                     |
| tag-ext                   | Tag support for resources: subnet, subnetpool, port, router |
| extra_dhcp_opt            | Neutron Extra DHCP opts                                     |
| standard-attr-revisions   | Resource revision numbers                                   |
| pagination                | Pagination support                                          |
| sorting                   | Sorting support                                             |
| security-group            | security-group                                              |
| dhcp_agent_scheduler      | DHCP Agent Scheduler                                        |
| router_availability_zone  | Router Availability Zone                                    |
| rbac-policies             | RBAC Policies                                               |
| standard-attr-description | standard-attr-description                                   |
| router                    | Neutron L3 Router                                           |
| allowed-address-pairs     | Allowed Address Pairs                                       |
| project-id                | project_id field enabled                                    |
| dvr                       | Distributed Virtual Router                                  |
+---------------------------+-------------------------------------------------------------+
```
It is indeed not there.
This is an Openstack Ocata installed on Centos 7

**Comment by bartv** [comment]
I have been looking at it and portsecurity is not an essential extension. You can have security groups without the portsecurity extension. However, it is the easiest manner to disable filtering for l2 an l3 network functions on openstack. It is similar to disabling source_dest_check on AWS.

I will provide a fix ASAP.

**Comment by bartv** [comment]
Could you try out the version in pull request #3 ?

**Comment by automate-solutions** [comment]
Nice! Thanks. I'll test that tomorrow morning.
I keep you posted.

**Comment by automate-solutions** [comment]
I still have the problem of versions supported for the openstack module.
A first, I specified 
`- openstack~=2.2.3` as a requirement in `project.yml`
and inmanta raised this warning:
```
no suitable version found for module openstack
WARNING  collecting reqs on project that has not been loaded completely
WARNING  collecting reqs on project that has not been loaded completely
WARNING  collecting reqs on project that has not been loaded completely
INFO     verifying project
WARNING  requirement openstack~=2.2.3 on module openstack not fullfilled, not at version 2.2.2
```
It got the `master` version instead of the `2.2.3.dev1498758188`

I then specified the full version required 
`- openstack==2.2.3.dev1498758188` with no more luck.
I then checked out directly in the openstack lib:
`git checkout issue/2`
And now I have that issue when trying to export:
```
[root@inmanta openstack-test]# inmanta -vvv export
INFO     Creating new virtual environment in ./.env
DEBUG    Starting compile
INFO     verifying project
WARNING  Empty deployment model.
could not find type std::AgentConfig in namespace std (reported in Construct(std::AgentConfig) (/tmp/tmpmqtlmkgr/openstack/model/_init.cf:42))
```

**Comment by bartv** [comment]
See my comment in issue #4 You can either use a dev rpm or apply the patch in the merge request yourself to the 2.1.0 version of the module.

**Comment by automate-solutions** [comment]
I Installed the dev version, deployed the openstack module and checked out to the `issue/2` branch.
When I run the deployment, I have that result:
```
[root@inmanta-dev openstack-dev]# inmanta -vvv deploy
INFO     Creating new virtual environment in ./.env
DEBUG    Storing state data in ./data/deploy
DEBUG    Creating directory ./data
DEBUG    Creating directory ./data/deploy
DEBUG    Starting mongodb on port 41181
DEBUG    Creating directory ./data/deploy/state
DEBUG    Creating directory ./data/deploy/logs
INFO     Starting server endpoint
INFO     Connected to mongodb database inmanta on localhost:41181
WARNING  Config name shared-secret not defined in section server
DEBUG    Cleaning server session DB
DEBUG    Started server on port 41979
DEBUG    Creating project deploy
DEBUG    Creating environment deploy in project ac299b92-aa42-4d04-b99a-9ef3e1a6c6fd
404 GET /parameter/ip_address?resource_id=openstack%3A%3AVirtualMachine%5Biaas_openstack%2Cname%3Dtesthost%5D (::1) 5.28ms
DEBUG    Starting purge_on_delete queries
DEBUG      All released versions: set()
DEBUG      Latest released version: None
DEBUG      Resource with purge_on_delete true: []
DEBUG      All resource in current version (1498807416): {'std::AgentConfig[internal,agentname=iaas_openstack]', 'openstack::HostPort[iaas_openstack,name=testhost_eth0]', 'openstack::VirtualMachine[iaas_openstack,name=testhost]'}
DEBUG      These resources are no longer present in current model: set()
DEBUG    Successfully stored version 1498807416
INFO     Export of model complete
INFO     Latest version for created environment is 1498807416
DEBUG    Agent(s) iaas_openstack, iaas_openstack, internal defined, adding them to autostart agent map
INFO     Environment setting autostart_agent_map changed. Restarting agents.
INFO     ['iaas_openstack', 'internal'] matches agents managed by server, ensuring it is started.
DEBUG    New session for agent iaas_openstack on inmanta-dev
DEBUG    New session for agent internal on inmanta-dev
DEBUG    set session 0888535c-5d65-11e7-a889-000c29543e51 as primary for agent iaas_openstack in env 01f405be-1560-49cb-ac3c-8c37e8aabdcf
DEBUG    set session 0888535c-5d65-11e7-a889-000c29543e51 as primary for agent internal in env 01f405be-1560-49cb-ac3c-8c37e8aabdcf
404 GET /resource?sid=0888535c-5d65-11e7-a889-000c29543e51&agent=iaas_openstack (::1) 23.25ms
DEBUG    Started new agent with PID 1612
INFO     ['iaas_openstack', 'internal'] matches agents managed by server, ensuring it is started.
std::AgentConfig[internal,agentname=iaas_openstack],v=1498807416 - deployed
[1 / 3]
[1 / 3]
[1 / 3]
openstack::VirtualMachine[iaas_openstack,name=testhost],v=1498807416 - deployed
openstack::HostPort[iaas_openstack,name=testhost_eth0],v=1498807416 - deployed
[3 / 3]
Deploy ready
DEBUG    expiring session 0888535c-5d65-11e7-a889-000c29543e51
```
And a grep doesn't return anything. Which seems to be good.
```
[root@inmanta-dev openstack-dev]# grep 'KeyError' data/deploy/logs/agent-01f405be-1560-49cb-ac3c-8c37e8aabdcf.err 
[root@inmanta-dev openstack-dev]# 
```

**Comment by bartv** [comment]
This was fixed in version 2.3.3 (just released)
